core/src/main/java/com/graphhopper/routing/ch/ManualPrepareContractionHierarchies.java

```diff
@@ -0,0 +1,123 @@
+        this.contractionOrder = contractionOrder;
+        return this;
+    }
+
+    public ManualPrepareContractionHierarchies setSeed(long seed) {
+        this.contractionOrder = createRandomContractionOrder(seed);
+        return this;
+    }
+
+    @Override
+    public String toString() {
+        return super.toString() + "|manual";
+    }
+
+    @Override
+    protected void runGraphContraction() {
+        setMaxLevelOnAllNodes();
+        contractNodes();
+    }
+
+    private void setMaxLevelOnAllNodes() {
+        int nodes = prepareGraph.getNodes();
+        for (int node = 0; node < nodes; node++) {
+            prepareGraph.setLevel(node, nodes);
+        }
+    }
+
+    private void contractNodes() {
+        if (contractionOrder.size() != prepareGraph.getNodes()) {
+            throw new IllegalArgumentException(String.format(Locale.ROOT,
+                    "contraction order size (%d) must be equal to number of nodes in graph (%d)",
+                    contractionOrder.size(), prepareGraph.getNodes()));
+        }
+        final EdgeFilter allFilter = DefaultEdgeFilter.allEdges(prepareWeighting.getFlagEncoder());
+        CHEdgeExplorer explorer = prepareGraph.createEdgeExplorer(allFilter);
+        CHEdgeExplorer discExplorer = prepareGraph.createEdgeExplorer(allFilter);
+        final int nodesToContract = contractionOrder.size();
+        final long logSize = 10;
+        StopWatch stopWatch = new StopWatch();
+        for (int i = 0; i < nodesToContract; ++i) {
+            stopWatch.start();
+            int node = contractionOrder.get(i);
+            // contract node
+            nodeContractor.contractNode(node);
+            long shortcutCount = nodeContractor.getAddedShortcutsCount();
+            prepareGraph.setLevel(node, i);
+
+            // disconnect neighbors
+            CHEdgeIterator iter = explorer.setBaseNode(node);
+            while (iter.next()) {
+                if (prepareGraph.getLevel(iter.getAdjNode()) != maxLevel)
+                    continue;
+                prepareGraph.disconnect(discExplorer, iter);
+            }
+            stopWatch.stop();
+            if (i % logSize == 0) {
+                logger.info(String.format("contracted %s / %s nodes, shortcuts: %s, last batch took: %.2f s, time per node: %.2f micros, %s",
+                        nf(i), nf(nodesToContract), nf(shortcutCount), stopWatch.getSeconds(),
+                        stopWatch.getNanos() / logSize * 1.e-3, nodeContractor.getStatisticsString()));
+            }
+        }
+    }
+
+    private List<Integer> createRandomContractionOrder(long seed) {
+        int nodes = prepareGraph.getNodes();
+        List<Integer> result = new ArrayList<>(nodes);
+        for (int i = 0; i < nodes; ++i) {
+            result.add(i);
+        }
+        // the shuffle method is the only reason we are using java.util.ArrayList instead of hppc IntArrayList here
```

Comment:
:) we should avoid ArrayList as it uses much more space for large node count and is probably also slower. (Here we could even use an int array)

Reply:
I agree, but I did not know of any easy way to shuffle an int array ? Right now this code is not used anywhere anyway and I removed it for now. This whole class will be reworked/moved when doing #1509.